### PR TITLE
#768 drafted logic for multi store support

### PR DESCRIPTION
**Pull request by ggarri**
## Fixed Issue(s)
#768 

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://github.com/consensys/orchestrate/blob/master/CHANGELOG.md).

### CODE REVIEW

**Review by 6l20 — APPROVED** [review]

File: pkg/types/api/account_request.go [hunk 1]
```diff
@@ -7,13 +7,15 @@ import (
 type CreateAccountRequest struct {
 	Alias      string            `json:"alias" validate:"omitempty" example:"personal-account" `
 	Chain      string            `json:"chain" validate:"omitempty" example:"besu"`
+	StoreID    string            `json:"storeID" validate:"omitempty" example:"qkmStoreID"`
```
Comment (darioAnongba) [inline pkg/types/api/account_request.go:10]: Why not just call it `Store json:"store"` ?

File: pkg/types/api/account_request.go [hunk 2]
```diff
@@ -7,13 +7,15 @@ import (
...
 	Attributes map[string]string `json:"attributes,omitempty"`
 }
 
 type ImportAccountRequest struct {
 	Alias      string            `json:"alias" validate:"omitempty" example:"personal-account"`
 	Chain      string            `json:"chain" validate:"omitempty" example:"quorum"`
 	PrivateKey hexutil.Bytes     `json:"privateKey" validate:"required" example:"0x66232652FDFFD802B7252A456DBD8F3ECC0352BBDE76C23B40AFE8AEBD714E2D" swaggertype:"string"`
+	StoreID    string            `json:"storeID" validate:"omitempty" example:"qkmStoreID"`
```
Comment (darioAnongba) [inline pkg/types/api/account_request.go:18]: Why not just call it Store json:"store" ?

File: pkg/types/entities/account.go [hunk 3]
```diff
@@ -14,6 +14,7 @@ type Account struct {
 	CompressedPublicKey hexutil.Bytes
 	TenantID            string
 	OwnerID             string
+	StoreID             string
```
Comment (darioAnongba) [inline pkg/types/entities/account.go:17]: Why not just call it Store ?

File: services/api/business/use-cases/accounts/create_account.go [hunk 4]
```diff
@@ -40,7 +40,7 @@ func NewCreateAccountUseCase(db store.DB, searchUC usecases.SearchAccountsUseCas
 		keyManagerClient: keyManagerClient,
 		fundAccountUC:    fundAccountUC,
 		logger:           log.NewLogger().SetComponent(createAccountComponent),
-		storeName:        qkm.GlobalStoreName(),
+		defaultStoreID:   qkm.GlobalStoreName(),
```
Comment (darioAnongba) [inline services/api/business/use-cases/accounts/create_account.go:43]: The use case should not be aware of these defaults, it is the parent that will check if a `store` is defined in the request and fill it with the default if it's empty. When reaching the use case, the account MUST have a store defined

File: services/api/business/use-cases/accounts/create_account.go [hunk 5]
```diff
@@ -64,10 +64,14 @@ func (uc *createAccountUseCase) Execute(ctx context.Context, account *entities.A
 		return nil, errors.AlreadyExistsError(errMsg).ExtendComponent(createAccountComponent)
 	}
 
+	if account.StoreID != "" {
```
Comment (darioAnongba) [inline services/api/business/use-cases/accounts/create_account.go:67]: The business layer should not be aware of these defaults, it is the service layer that will check if a `store` is defined in the request and fill it with the default if it's empty. When reaching the use case, the account MUST have a store defined

File: services/api/store/postgres/migrations/22_add_account_storeID.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package migrations
+
+import (
+	"github.com/go-pg/migrations/v7"
+	log "github.com/sirupsen/logrus"
+)
+
+func upgradeAccountStoreID(db migrations.DB) error {
```
Comment (darioAnongba) [inline services/api/store/postgres/migrations/22_add_account_storeID.go:8]: If you already had accounts before v21.12. It means that you should first run the migrations, then import then accounts into the store. Are we missing something in the `imports` script?

File: services/api/service/formatters/accounts.go [hunk 7]
```diff
@@ -14,13 +14,15 @@ func FormatCreateAccountRequest(req *api.CreateAccountRequest) *entities.Account
 	return &entities.Account{
 		Alias:      req.Alias,
 		Attributes: req.Attributes,
+		StoreID:    req.StoreID,
```
Comment (darioAnongba) [inline services/api/service/formatters/accounts.go:17]: this is where the default value should be set if empty in the request.

File: services/api/service/formatters/accounts.go [hunk 8]
```diff
@@ -14,13 +14,15 @@ func FormatCreateAccountRequest(req *api.CreateAccountRequest) *entities.Account
...
 	}
 }
 
 func FormatImportAccountRequest(req *api.ImportAccountRequest) *entities.Account {
 	return &entities.Account{
 		Alias:      req.Alias,
 		Attributes: req.Attributes,
+		StoreID:    req.StoreID,
```
Comment (darioAnongba) [inline services/api/service/formatters/accounts.go:25]: this is where the default value should be set if empty in the request.

File: pkg/types/api/account_request.go (hunk 1 above)
Comment (ggarri) [reply pkg/types/api/account_request.go:10]: Discussed an defined on storeID

File: services/api/business/use-cases/accounts/create_account.go (hunk 4 above)
Comment (ggarri) [reply services/api/business/use-cases/accounts/create_account.go]: After an Slack discussion it was decided to move on with this approach and create a tech debt ticket to apply same strategy for other default fields in the codebase such as priority or retryInterval

File: services/api/store/postgres/migrations/22_add_account_storeID.go (hunk 6 above)
Comment (ggarri) [reply services/api/store/postgres/migrations/22_add_account_storeID.go:8]: Accounts having an empty StoreID will be using default StoreID.

File: services/api/business/use-cases/accounts/create_account.go (hunk 5 above)
Comment (ggarri) [reply services/api/business/use-cases/accounts/create_account.go]: Removed

File: services/api/service/formatters/accounts.go (hunk 7 above)
Comment (ggarri) [reply services/api/service/formatters/accounts.go:17]: Applied change

File: services/api/service/formatters/accounts.go (hunk 8 above)
Comment (ggarri) [reply services/api/service/formatters/accounts.go:31]: applied

**Pull request merged**